src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,279 @@
+from pants.build_graph.address import Address
+from pants.engine.addressable import BuildFileAddresses
+from pants.engine.fs import (
+  Digest,
+  DirectoriesToMerge,
+  DirectoryWithPrefixToAdd,
+  FileContent,
+  FilesContent,
+  InputFilesContent,
+)
+from pants.engine.goal import Goal, GoalSubsystem
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.graph import HydratedTarget, TransitiveHydratedTargets, HydratedTargets
+from pants.engine.rules import goal_rule, rule, subsystem_rule
+from pants.engine.selectors import Get, MultiGet
+from pants.rules.core.strip_source_root import SourceRootStrippedSources
+from pants.rules.core.test import AddressAndTestResults, PytestCoverageReport
+from pants.source.source_root import SourceRootConfig
+
+
+class CoverageToolBase(PythonToolBase):
+  options_scope = 'pytest-coverage'
+  default_version = 'coverage==5.0.0'
+  default_entry_point = 'coverage'
+  default_interpreter_constraints = ["CPython>=3.6"]
+
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register(
+      '--output-path',
+      type=str,
+      default='coverage/python',
+      help='Path to write pytest coverage report to. Must be relative to build root.',
+    )
+    register(
+      '--report',
+      type=str,
+      choices=('xml', 'html'),
+      default='html',
+      help='Which coverage reports to emit.',
+    )
+
+
+@dataclass(frozen=True)
+class CoverageSetup:
+  requirements_pex: Pex
+  filename: str
+
+
+@rule
+async def setup_coverage(
+  coverage: CoverageToolBase,
+) -> CoverageSetup:
+  plugin_file_digest: Digest = await Get[Digest](InputFilesContent, get_coverage_plugin_input())
+  output_pex_filename = "coverage.pex"
+  requirements_pex = await Get[Pex](
+    CreatePex(
+      output_filename=output_pex_filename,
+      requirements=PexRequirements(requirements=tuple(coverage.get_requirement_specs())),
+      interpreter_constraints=PexInterpreterConstraints(
+        constraint_set=tuple(coverage.default_interpreter_constraints)
+      ),
+      entry_point=coverage.get_entry_point(),
+      input_files_digest=plugin_file_digest,
+    )
+  )
+  return CoverageSetup(requirements_pex, output_pex_filename)
+
+
+@dataclass(frozen=True)
+class MergedCoverageData:
+  coverage_data: Digest
+
+@rule(name="Merge coverage reports")
+async def merge_coverage_reports(
+  test_results: AddressAndTestResults,
+  transitive_targets: TransitiveHydratedTargets,
+  python_setup: PythonSetup,
+  coverage_setup: CoverageSetup,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment,
+) -> MergedCoverageData:
+  """Takes all python test results and merges their coverage data into a single sql file."""
+
+  coveragerc_digest = await Get[Digest](InputFilesContent, get_coveragerc_input(DEFAULT_COVERAGE_CONFIG))
+
+  coverage_directory_digests: Tuple[Digest, ...] = await MultiGet(
```

Comment:
Same with trying again please to see if MyPy can infer what this is. It correctly infers in most places with `MultiGet`.